presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java

```diff
@@ -416,8 +416,11 @@ public void addPreparedStatement(String key, String value)
         requireNonNull(key, "key is null");
         requireNonNull(value, "value is null");
 
+        if (session.getPreparedStatements().containsKey(key)) {
+            throw new PrestoException(ALREADY_EXISTS, "Prepared statement already exists: " + key);
+        }
         String previousValue = addedPreparedStatements.putIfAbsent(key, value);
-        if (previousValue != null || session.getPreparedStatements().containsKey(key)) {
+        if (previousValue != null) {
```

Comment:
I'm a bit confused by the question, but I'll try to answer.  I

If the key does not already exist as part of the session, then the first if statement will get passed by.

The second if statement checks the updates that are part of the current transaction.  Those get put in addedPreparedStatements.  You expect that  addedPreparedStatements does not yet have key with the name of your prepared statement, so previousValue will be null.  However, if you already in the transaction tried to update the prepared statement, previousValue will not be null, and you'll hit this presto exception.  There is no way to distinguish between a key that was already in the addedPreparedStatements table but had a value of null, and a key that was not there and therefore null was returned.  But it doesn't matter because Prepared Statements cannot be null.